Source/WebCore/platform/graphics/MediaResourceSniffer.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "MediaResourceSniffer.h"
+
+#include "MIMESniffer.h"
+#include "ResourceRequest.h"
+
+namespace WebCore {
+
+RefPtr<MediaResourceSniffer> MediaResourceSniffer::create(PlatformMediaResourceLoader& loader, ResourceRequest&& request, size_t maxSize)
+{
+    if (maxSize != size_t(-1))
+        request.addHTTPHeaderField(HTTPHeaderName::Range, makeString("bytes=", 0, '-', maxSize));
+    auto resource = loader.requestResource(WTFMove(request), PlatformMediaResourceLoader::LoadOption::DisallowCaching);
+    if (!resource)
+        return nullptr;
+    auto sniffer = adoptRef(*new MediaResourceSniffer { Ref { *resource }, maxSize });
+    auto result = sniffer.copyRef();
+    resource->setClient(WTFMove(sniffer));
+    return result;
+}
+
+MediaResourceSniffer::MediaResourceSniffer(Ref<PlatformMediaResource>&& resource, size_t maxSize)
+    : m_resource(WTFMove(resource))
+    , m_maxSize(maxSize)
+{
+}
+
+MediaResourceSniffer::~MediaResourceSniffer()
+{
+    cancel();
+}
+
+void MediaResourceSniffer::cancel()
+{
+    if (auto resource = WTFMove(m_resource))
+        resource->shutdown();
+    if (!m_producer)
+        m_producer.reject(PlatformMediaError::Cancelled);
+    m_content.reset();
+}
+
+Ref<MediaResourceSniffer::Promise> MediaResourceSniffer::promise() const
+{
+    return m_producer.promise();
+}
+
+void MediaResourceSniffer::dataReceived(PlatformMediaResource&, const SharedBuffer& buffer)
+{
+    auto mimeType = MIMESniffer::getMIMETypeFromContent({ buffer.data(), buffer.size() });
```

Comment:
This seems incorrect; it would only work on the first data that's received. Shouldn't the `m_content.append(buffer)` step happen first, and _then_ the sniff?

Reply:
good point, I believe that in practice it will always work on the first packet received, as it's unlikely we will receive less than the 1485 bytes we requested.

Reply:
`{ buffer.data(), buffer.size() }` -> `buffer.dataAsSpanForContiguousData()`. It is a horrible name so I intend to rename soon but still.